lib/clean_test.py

```diff
@@ -36,3 +38,131 @@ def test_remove_future_dates(self):
+                            "first_name": "thomas",
+                            "last_name": "ferguson",
+                        },
+                        {
+                            "uid": "a5f3c016d4c3373aa74dc15c0638362e",
+                            "first_name": "thomas",
+                            "last_name": "ferguso",
+                        },
+                        {
+                            "uid": "233418697784e972144523ad8cc4ed9",
+                            "first_name": "arthur",
+                            "middle_name": "hesse",
+                            "last_name": "schopenhaur",
+                        },
+                        {
+                            "uid": "d59c942cdd7e211bf7c76f20501d657c",
+                            "first_name": "arthur",
+                            "middle_name": "h",
+                            "last_name": "schopenhaur",
+                        },
+                    ]
+                ),
+                clusters=[
+                    (
+                        "db3d392b404754b2d4a127ca0922d2f8",
+                        "1e65c0807675ee3f25f6a6bf25eb121b",
+                    ),
+                    (
+                        "92d3d9c79a3eb44ece5c83e62e55b91d",
+                        "a5f3c016d4c3373aa74dc15c0638362e",
+                    ),
+                    (
+                        "233418697784e972144523ad8cc4ed9",
+                        "d59c942cdd7e211bf7c76f20501d657c",
+                    ),
+                ],
+            ).fillna(""),
+            pd.DataFrame(
+                [
+                    {
+                        "uid": "db3d392b404754b2d4a127ca0922d2f8",
+                        "first_name": "",
+                        "last_name": "",
+                    },
+                    {
+                        "uid": "db3d392b404754b2d4a127ca0922d2f8",
+                        "first_name": "patrick",
+                        "last_name": "peterman",
+                    },
+                    {
+                        "uid": "92d3d9c79a3eb44ece5c83e62e55b91d",
+                        "first_name": "thomas",
+                        "last_name": "ferguson",
+                    },
+                    {
+                        "uid": "92d3d9c79a3eb44ece5c83e62e55b91d",
+                        "first_name": "",
+                        "last_name": "",
+                    },
+                    {
+                        "uid": "233418697784e972144523ad8cc4ed9",
+                        "first_name": "arthur",
+                        "middle_name": "hesse",
+                        "last_name": "schopenhaur",
+                    },
+                    {
+                        "uid": "233418697784e972144523ad8cc4ed9",
+                        "first_name": "",
+                        "middle_name": "",
+                        "last_name": "",
+                    },
```

Comment:
It's strange that the name columns of non-canonical officers are empty. I don't think that's what we want, rather all columns including `uid` should be the same as the canonical one. Can you add some non-name columns too so we can assert that this function leaves other columns as is?

Reply:
I also thought that was strange. I've added non-name columns to assert that this function leaves other columns as is. The code that you originally wrote (pasted below) also removed names and only replaced `uids`, which made me, believe that this was the intended output. Should this then be re-written to include the` name` columns in addition to the `uid` columns?

```
def canonicalize_names(
        df: pd.DataFrame, clusters: list[tuple], uid_column: str = 'uid',
        first_name_column: str = 'first_name', last_name_column: str = 'last_name'
) -> pd.DataFrame:
    for cluster in clusters:
        uid, first_name, last_name = None, '', ''
        for idx in cluster:
            row = df.loc[df[uid_column] == idx]
            if (
                uid is None
                or len(row[first_name_column]) > len(first_name)
                or (len(row[first_name_column]) == len(first_name) and len(row[last_name_column]) > len(last_name))
            ):
                uid, first_name, last_name = idx, row[first_name_column], row[last_name_column]
        df.loc[df[uid_column].isin(cluster), uid_column] = uid
        df.loc[df[uid_column] == uid, first_name_column] = first_name
        df.loc[df[uid_column] == uid, last_name_column] = last_name
    return df
```

Reply:
Yes, it will be helpful in this instance to write the expected output first and then modify the code to match the output. Congrats! You have just discovered TDD. :)

Don't worry about my other suggestion, just focus on making the function correct for now.